index.js

```diff
@@ -16,65 +20,67 @@ util.inherits(CpFileError, NestedError);
 
 CpFileError.prototype.name = 'CpFileError';
 
-module.exports = function (src, dest, opts, cb) {
+module.exports = function (src, dest, opts) {
 	if (!src || !dest) {
 		throw new CpFileError('`src` and `dest` required');
 	}
 
 	if (typeof opts !== 'object') {
-		cb = opts;
 		opts = {};
 	}
 
-	cb = onetime(cb || function () {});
 	opts = objectAssign({overwrite: true}, opts);
 
-	var read = fs.createReadStream(src);
-	var readListener = onetime(startWrite);
+	return new Promise(function startRead(resolve, reject) {
+		var read = fs.createReadStream(src);
 
-	read.on('error', function (err) {
-		cb(new CpFileError('cannot read from `' + src + '`: ' + err.message, err));
-	});
-	read.on('readable', readListener);
-	read.on('end', readListener);
-
-	function startWrite() {
-		mkdirp(path.dirname(dest), function (err) {
-			if (err && err.code !== 'EEXIST') {
-				cb(new CpFileError('cannot create directory `' + path.dirname(dest) + '`: ' + err.message, err));
-				return;
+		read.on('error', function (err) {
+			reject(new CpFileError('cannot read from `' + src + '`: ' + err.message, err));
+		});
+		read.on('readable', function () {
+			resolve(read);
+		});
+		read.on('end', function () {
+			resolve(read);
```

Comment:
I think the second one will be omitted as the promise is already resolved. But I guess this could be written more nicely with [get-stream](https://github.com/sindresorhus/get-stream).